### ISSUE TITLE
update_rate causing NaNs in rrd file

### ISSUE THREAD

**Issue by kellerkindt** [munin-monitoring/munin#711]
### == issue explanation ==

The update_rate causes values not to be inserted into the rrd file, if the time diff
since the previous lookup is to low (see second debug output `2016/06/19 23:01:02 last: [1466370022,526128], now: [1466370062, 635050], age: 40.108922`) )(Sometimes even ~55s seem to be to low).

Removing the `update_rate 60`from the configuration removes this check, which will insert all collected values to be inserted in the rrd file if it already exists, but the rrd-file cannot be created with 'Script v2' (`2016/06/20 00:14:02 [ERROR] Unable to create '/var/lib/munin/hebeanlage/hebeanlage-hebeanlage_warning-warning-g.rrd': Invalid step: must be >= 1`) 

'Script v3' creates for each request, event if it does not get into the rrd-file a valid line in the tmp file and therefore proves that the munin-script works fine (beside from the debug outputs, that both receive the value)
### == How to reproduce ==

Add 'Script v1' to your munin-node plugins folder
### == Workaround ==

Run 'Script v1' once, so the rrd-file gets created,
then replace it with 'Script v2'
### == preferred fix ==

Having something like `graph_data_rate 60` which is like `update_rate 60` but valid only for the rrd file
### == munin version and system environment ==

`debian 8.5`
`munin 2.0.25-1`
### == Script v1 ==

``` bash
#!/bin/bash

case $1 in
   config)
        cat <<'EOM'
graph_data_size custom 1m for 1d, 1m for 1w, 5m for 1t, 15m for 1y
update_rate 60
graph_title Hebeanlage Warnungstatus
graph_vlabel warning
graph_category haus
warning.draw AREA
warning.label Warning status
warning.colour FFA500
warning.critical :0
EOM
        exit 0;;


esac

echo "warning.value $(/usr/local/bin/i2c_hebeanlage /dev/i2c-1 8 gw)"
```
### == Script v2 ==

``` bash
#!/bin/bash

case $1 in
   config)
        cat <<'EOM'
graph_data_size custom 1m for 1d, 1m for 1w, 5m for 1t, 15m for 1y
graph_title Hebeanlage Warnungstatus
graph_vlabel warning
graph_category haus
warning.draw AREA
warning.label Warning status
warning.colour FFA500
warning.critical :0
EOM
        exit 0;;

esac

echo "warning.value $(/usr/local/bin/i2c_hebeanlage /dev/i2c-1 8 gw)"
```
### == Script v3 ==

``` bash
#!/bin/bash

case $1 in
   config)
        cat <<'EOM'
graph_data_size custom 1m for 1d, 1m for 1w, 5m for 1t, 15m for 1y
update_rate 60
graph_title Hebeanlage Warnungstatus
graph_vlabel warning
graph_category haus
warning.draw AREA
warning.label Warning status
warning.colour FFA500
warning.critical :0
EOM
        exit 0;;

esac

echo "warning.value $(/usr/local/bin/i2c_hebeanlage /dev/i2c-1 8 gw)" | tee -a /tmp/hebeanlage_warning.value.log
```
### == rrd-file ==

```
                        <!-- 2016-06-19 22:38:00 CEST / 1466368680 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:39:00 CEST / 1466368740 --> <row><v>1.0000000000e+00</v></row>
                        <!-- 2016-06-19 22:40:00 CEST / 1466368800 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:41:00 CEST / 1466368860 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:42:00 CEST / 1466368920 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:43:00 CEST / 1466368980 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:44:00 CEST / 1466369040 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:45:00 CEST / 1466369100 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:46:00 CEST / 1466369160 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:47:00 CEST / 1466369220 --> <row><v>1.0000000000e+00</v></row>
                        <!-- 2016-06-19 22:48:00 CEST / 1466369280 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:49:00 CEST / 1466369340 --> <row><v>1.0000000000e+00</v></row>
                        <!-- 2016-06-19 22:50:00 CEST / 1466369400 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:51:00 CEST / 1466369460 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:52:00 CEST / 1466369520 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:53:00 CEST / 1466369580 --> <row><v>1.0000000000e+00</v></row>
                        <!-- 2016-06-19 22:54:00 CEST / 1466369640 --> <row><v>1.0000000000e+00</v></row>
                        <!-- 2016-06-19 22:55:00 CEST / 1466369700 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:56:00 CEST / 1466369760 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:57:00 CEST / 1466369820 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 22:58:00 CEST / 1466369880 --> <row><v>1.0000000000e+00</v></row>
                        <!-- 2016-06-19 22:59:00 CEST / 1466369940 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 23:00:00 CEST / 1466370000 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 23:01:00 CEST / 1466370060 --> <row><v>NaN</v></row>
                        <!-- 2016-06-19 23:02:00 CEST / 1466370120 --> <row><v>1.0000000000e+00</v></row>
```
### == value gets inserted in rrd file ==

```
2016/06/19 23:02:02 [DEBUG] Reading from socket: "graph_data_size custom 1m for 1d, 1m for 1w, 5m for 1t, 15m for 1y\nupdate_rate 30\ngraph_title Hebeanlage Warnungstatus\ngraph_vlabel warning\ngraph_category haus\nwarning.draw AREA\nwarning.label Warning status\nwarning.colour FFA500\nwarning.critical :0"
.
2016/06/19 23:02:02 [DEBUG] config: 0.114633 sec for 'hebeanlage_warning' on hebeanlage/hebeanlage.<server-addr>/4949
2016/06/19 23:02:02 [DEBUG] Now parsing config output from plugin hebeanlage_warning on hebeanlage
2016/06/19 23:02:02 [DEBUG] update_rate 30 for hebeanlage_warning on hebeanlage/hebeanlage.<server-addr>:4949
2016/06/19 23:02:02 is_fresh_enough asked for hebeanlage_warning with a rate of 30
2016/06/19 23:02:02 last_updated{hebeanlage_warning}: 1466370062 635050
2016/06/19 23:02:02 last: [1466370062,635050], now: [1466370122, 210740], age: 59.57569
2016/06/19 23:02:02 is_fresh_enough  0
2016/06/19 23:02:02 new value: 1466370122 210740
2016/06/19 23:02:02 [DEBUG] No service data for hebeanlage_warning, fetching it
2016/06/19 23:02:02 [DEBUG] Writing to socket: "fetch hebeanlage_warning
".
2016/06/19 23:02:02 [DEBUG] data: 0.126496 sec for 'hebeanlage_warning' on hebeanlage/hebeanlage.<server-addr>/4949
2016/06/19 23:02:02 [DEBUG] Now parsing fetch output from plugin hebeanlage_warning on hebeanlage/hebeanlage.<server-addr>:4949
2016/06/19 23:02:02 [FETCH from hebeanlage_warning] warning.value 1
2016/06/19 23:02:02 [FETCH from hebeanlage_warning] Storing 1 in warning
2016/06/19 23:02:02 [DEBUG] asking for a rrd of size : custom 1m for 1d, 1m for 1w, 5m for 1t, 15m for 1y
2016/06/19 23:02:02 [DEBUG] rrd filename: /var/lib/munin/hebeanlage/hebeanlage-hebeanlage_warning-warning-g.rrd
2016/06/19 23:02:02 [DEBUG] Updating /var/lib/munin/hebeanlage/hebeanlage-hebeanlage_warning-warning-g.rrd with 1466370122:1
2016/06/19 23:02:02 [DEBUG] set_spoolfetch_timestamp(1466370122)
2016/06/19 23:02:02 [DEBUG] Writing to socket: "quit
".
2016/06/19 23:02:02 [DEBUG] quit: 0.00012 sec on hebeanlage/hebeanlage.<server-addr>/4949
2016/06/19 23:02:02 [DEBUG] Everything went smoothly.
2016/06/19 23:02:02 [DEBUG] Writing state for hebeanlage-hebeanlage in /var/lib/munin/state-hebeanlage-hebeanlage.storable
2016/06/19 23:02:02 [DEBUG] about to write '/var/lib/munin/state-hebeanlage-hebeanlage.storable'
2016/06/19 23:02:02 [INFO]: Munin-update finished for node hebeanlage;hebeanlage (1.12 sec)
2016/06/19 23:02:02 [DEBUG] Lock /var/run/munin/munin-hebeanlage2-hebeanlage2.lock already exists, checking process
2016/06/19 23:02:02 [DEBUG] Lock contained pid '3380'
```
### == value does not get inserted in rrd file ==

```
2016/06/19 23:01:02 [DEBUG] Reading from socket: "graph_data_size custom 1m for 1d, 1m for 1w, 5m for 1t, 15m for 1y\nupdate_rate 30\ngraph_title Hebeanlage Warnungstatus\ngraph_vlabel warning\ngraph_category haus\nwarning.draw AREA\nwarning.label Warning status\nwarning.colour FFA500\nwarning.critical :0"
.
2016/06/19 23:01:02 [DEBUG] config: 0.118532 sec for 'hebeanlage_warning' on hebeanlage/hebeanlage.<server-addr>/4949
2016/06/19 23:01:02 [DEBUG] Now parsing config output from plugin hebeanlage_warning on hebeanlage
2016/06/19 23:01:02 [DEBUG] update_rate 30 for hebeanlage_warning on hebeanlage/hebeanlage.<server-addr>:4949
2016/06/19 23:01:02 is_fresh_enough asked for hebeanlage_warning with a rate of 30
2016/06/19 23:01:02 last_updated{hebeanlage_warning}: 1466370022 526128
2016/06/19 23:01:02 last: [1466370022,526128], now: [1466370062, 635050], age: 40.108922
2016/06/19 23:01:02 is_fresh_enough  0
2016/06/19 23:01:02 new value: 1466370062 635050
2016/06/19 23:01:02 [DEBUG] No service data for hebeanlage_warning, fetching it
2016/06/19 23:01:02 [DEBUG] Writing to socket: "fetch hebeanlage_warning
".
2016/06/19 23:01:02 [DEBUG] data: 0.12606 sec for 'hebeanlage_warning' on hebeanlage/hebeanlage.<server-addr>/4949
2016/06/19 23:01:02 [DEBUG] Now parsing fetch output from plugin hebeanlage_warning on hebeanlage/hebeanlage.<server-addr>:4949
2016/06/19 23:01:02 [FETCH from hebeanlage_warning] warning.value 1
2016/06/19 23:01:02 [FETCH from hebeanlage_warning] Storing 1 in warning
2016/06/19 23:01:02 [DEBUG] asking for a rrd of size : custom 1m for 1d, 1m for 1w, 5m for 1t, 15m for 1y
2016/06/19 23:01:02 [DEBUG] rrd filename: /var/lib/munin/hebeanlage/hebeanlage-hebeanlage_warning-warning-g.rrd
2016/06/19 23:01:02 [DEBUG] Updating /var/lib/munin/hebeanlage/hebeanlage-hebeanlage_warning-warning-g.rrd with 1466370062:1
2016/06/19 23:01:02 [DEBUG] set_spoolfetch_timestamp(1466370062)
2016/06/19 23:01:02 [DEBUG] Writing to socket: "quit
".
2016/06/19 23:01:02 [DEBUG] quit: 0.000161 sec on hebeanlage/hebeanlage.<server-addr>/4949
2016/06/19 23:01:02 [DEBUG] Everything went smoothly.
2016/06/19 23:01:02 [DEBUG] Writing state for hebeanlage-hebeanlage in /var/lib/munin/state-hebeanlage-hebeanlage.storable
2016/06/19 23:01:02 [DEBUG] about to write '/var/lib/munin/state-hebeanlage-hebeanlage.storable'
2016/06/19 23:01:02 [INFO]: Munin-update finished for node hebeanlage;hebeanlage (1.14 sec)
2016/06/19 23:01:02 [DEBUG] Lock /var/run/munin/munin-hebeanlage2-hebeanlage2.lock already exists, checking process
2016/06/19 23:01:02 [DEBUG] Lock contained pid '2992'
```

**Comment by kellerkindt** [comment]
Another graph but having the same issue will then look like this:

![bildschirmfoto vom 2016-06-20 00 25 30](https://cloud.githubusercontent.com/assets/1133787/16180176/ca774898-367d-11e6-94fb-4aee4b5ccbc9.png)

**Comment by kellerkindt** [comment]
This issue still exists.
```
munin: 2.0.33-1 
munin-node: 2.0.33-1
debian 9.4
```

Had to change update_rate to 30 for the first call to the script (where the rrd file is created), otherwise the graph has holes like that:
![bildschirmfoto vom 2018-03-18 14-45-42](https://user-images.githubusercontent.com/1133787/37567123-77e27cfc-2ac2-11e8-88ef-fd9b6129bbbb.png)

After the workaround and setting the update_rate to 60 again, the new graph looks fine:

![bildschirmfoto vom 2018-03-18 15-33-27](https://user-images.githubusercontent.com/1133787/37567128-9202c128-2ac2-11e8-8b80-c0b6c6d32526.png)

**Comment by sumpfralle** [comment]
Indeed the check for an elapsed `update_rate` was too strict before.

I fixed this now in b29b368013f19aff6f458d8c28ab314f8727141d.

Thanks for your report.
